src/main/ccl/cclutmock_table.inc

```diff
@@ -397,6 +405,12 @@ subroutine public::cclutCreateMockTable(cclutTableName)
     declare cclutParserText = vc with protect, noconstant("")
     declare cclutColumnName = vc with protect, noconstant("")
 
+    ;Verify uesr is DBA or the session is authenticated
```

Comment:
Moving the discussion from #31 to here.
 
>feckertson>>
>Inspired by this PR, I picked work on #29 back up.  
In the process, I found that perhaps cclutCreateMockTable does not work as one would like. It does not return a value that the consumer could check and it does not stop further execution when there is a problem. It just logs an error message and trudges on.

>It definitely should not do that in the case of an unauthenticated group1 user. Otherwise there will just be scads of messages saying "The CCL session must be authenticated to execute this test" followed by scads of messages saying "cclutAssertMockTableIsCreated(XXXX) - Table has not been created." followed by whatever assert failures occur because of those problems. The resolution is to go to exit_script rather than return in that case.

>Do we want it to do the same for the other kinds of problems that can occur?


>cernertrevor>>
>Do you have an example of this? I know you mentioned it in the conversation for #29 but I wasn't really sure what you were talking about, and since it didn't impact the review, I set it aside.

>I've set up the scenario I believe you're discussing (user has access to the back-end but no authenticated CCL session, reqinfo->updt_id returns 0.0, user is group1), but the create table commands appear to work fine (the messages you indicate don't appear). I do get the drop table issues (which is what #29 was originally for), but I'm not clear on this other issue.


The latest push for #29 terminates the test case with an error. In the case where a group1 user is trying mock a table in an unauthenticated session, it makes sense to fail fast. The question is whether it makes sense to fail fast for any other types of problems. Since there is a fail fast parameter, I tend to think no, but..... it could certainly make sense to fail fast in the current test (not the entire test case), if there were a way to pull that off.

Reply:
> The latest push for #29 terminates the test case with an error. In the case where a group1 user is trying mock a table in an unauthenticated session, **it makes sense to fail fast**.

The bolded part does not make sense to me.  Why does it make sense to fail fast?  The claim is that there is some issue creating tables:

> Otherwise there will just be scads of messages saying "The CCL session must be authenticated to execute this test" **followed by scads of messages saying "cclutAssertMockTableIsCreated(XXXX) - Table has not been created." followed by whatever assert failures occur because of those problems.**

This is not what I'm seeing (hence why I asked for an example).  

I do need to correct myself on one aspect though.  My statement here:

> I do get the drop table issues (which is what #29 was originally for), but I'm not clear on this other issue.

is wrong.  The drop table issue is the tdbexecute failing, not the duplicate mock table name that #29 was originally fixing.  Is the tdbexecute possibly what you're referencing?  That would make more sense than your original claim, but it's still misleading to suggest that the error is with creating the table, when the error is with dropping it.  I also disagree that exiting the entire test case is the correct approach.  It correctly generates the error that the table still exists and requires clean-up.  It's visible in the error section of the output report, and it has no bearing on other tests (if it's still possible for us to run other tests, I think we should instead of stopping execution by going to exit_script).

Reply:
You are referring to an error message that says "failed to drop mock table X" which happens because CCL Unit tries to drop a table but cannot. This code prevents the mock table from even being create in the first place and the message says "you need to authenticate in order to run this test case". 

CCL Unit should not be allowed to create a mock table for a group1 user in an unauthenticated session because CCL Unit will not be able to drop that table. 

With this change, the mock table does not get created because of an environmental problem, not because of a problem with the test. The user needs to fix that problem by authenticating (or using a test name pattern to skip the problematic tests).  

It seems rather pointless to continue any further testing once this situation has been identified and doing so may in fact make it more difficult for the user to understand what is really the matter not to mention that the messaging looks really amateurish in that case. To see this, just replace the go to exit_script with return in this code and then use cclut to execute a test case which passes and which mocks multiple tables (even better if it uses multiple tests to do so) and do so as a group1 user in an unauthenticated session.

Reply:
> You are referring to an error message that says "failed to drop mock table X" which happens because CCL Unit tries to drop a table but cannot. This code prevents the mock table from even being create in the first place and the message says "you need to authenticate in order to run this test case".

Actually, what I'm doing is making assumptions because I'm not able to get you to provide an example or confirm what you're talking about.

I'll reiterate it.
The original claim is that there is some issue creating tables:

> Otherwise there will just be scads of messages saying "The CCL session must be authenticated to execute this test" followed by scads of messages saying "cclutAssertMockTableIsCreated(XXXX) - Table has not been created." followed by whatever assert failures occur because of those problems.

Because I haven't received confirmation from you on the issue, I've chosen to make assumptions about what I _think_ you're talking about.  That leads me to believe you're talking about the issue where tdbexecute fails to drop a table due to the unauthenticated session.  This should not cause any assertion failures (assuming the original fix for #29 is present), nor should it cause any errors on table creation (both of which were implied by your message above).  In fact, the only reason I originally created it as an error (on my original PR) was to get the user's attention.  In reality, I'd only consider it a warning, because everything about the test works and the test passes, but in such situations, I doubt users would bother to read a warning, leading to leftover tables in the domain.

> With this change, the mock table does not get created because of an environmental problem, not because of a problem with the test. The user needs to fix that problem by authenticating (or using a test name pattern to skip the problematic tests).

The problem is, they may not know what all tests are problematic because we instantly failed fast.  They'll know the first one, but not any others.  In this instance, I'm a fan of giving more information rather than less.  If the proposal was to have it fail fast whenever they execute from cclut_ff (somewhat alluded to in your earlier comment), I also think that would be appropriate.

> It seems rather pointless to continue any further testing once this situation has been identified and doing so may in fact make it more difficult for the user to understand what is really the matter not to mention that the messaging looks really amateurish in that case. To see this, just replace the go to exit_script with return in this code and then use cclut to execute a test case which passes and which mocks multiple tables (even better if it uses multiple tests to do so) and do so as a group1 user in an unauthenticated session.

I would think under such logic, continuing to execute any test once an assertion failure or error has been identified is also pointless.  They have to fix the assertion failure anyway (or exclude it from running), so why bother wasting time running other tests?  

The reason is the user has an option of failing fast or not, and if they choose not, we can still run the test, and the user can still see if there were any assertion failures (meaning the table dropping was the only issue, or if they have other issues they also need to resolve).

If the messaging is "amateurish" (though I disagree with this as well; I ran the scenario you suggested and it seems fine to me), is it just a matter of changing the text?  Note that if you want to have a different message when the user is unauthenticated group1, that's also fine with me.

I'm not really sure what you're looking for from me.  Technically, I had already approved #29 (even though I disagree with the latest commit), and you're the maintainer, so I can't stop you from making this change.  If you just want an answer to your original question,

> Do we want it to do the same for the other kinds of problems that can occur?

Well, _I_ don't, but I also don't want it to happen here either, so that may not be a suitable answer for you.

Reply:
I am talking about the code we are currently commenting on and more specifically line 411. If you want to see what I am talking about. Try running it. Also try changing line 411 to say "return" rather than "go to exit_script". Any test case that creates several mock tables will make a suitable example.

Do you agree that the framework should not create tables when it knows it cannot remove them? 
Or do you think it should go ahead and create whatever tables the consumer asked for and then message out "Unable to drop table.  Manual cleanup may be required.  Mock table name is X" for every single one of them and leave it to the consumer to do said manual cleanup?


To provide some context why I thought you might have some idea what I'm taking about, we have my response 
>Yes. I had done that manually, but forgot I hadn't made a unit test for it (essentially the example from the issue).

>Another thing that should happen is forcing the session to be authenticated (for group 1 users). It doesn't really need to occur unless cclutCreateMockTable gets called.

to your last comment on this PR
>Actually, I just approved it, but one quick thought I had. Should we update one of the unit tests to ensure that the number continues counting up even after cclutRemoveAllMocks is called and cclutDefineMockTable is called again?

Reply:
> I am talking about the code we are currently commenting on and more specifically line 411. If you want to see what I am talking about. Try running it. Also try changing line 411 to say "return" rather than "go to exit_script". Any test case that creates several mock tables will make a suitable example.

I'm talking about the original comment (which spawned the code change that's being commented on).  I'm attempting to be diplomatic in this exchange, but if it helps for me to call it out specifically, I can.

> In the process, I found that perhaps cclutCreateMockTable does not work as one would like.

This is an opinion, but is necessary to understand the comments below.

> It does not return a value that the consumer could check and it does not stop further execution when there is a problem. **It just logs an error message and trudges on.**

This is false.  No error message is logged.  It "trudges on" because the creation succeeded.  Creating a table does not fail in an unauthenticated group1 session.  If you have an example to prove me wrong, I will gladly accept it.  I asked for one and was not provided with one.  I did set up the scenario (which demonstrated that the table creation does not fail).  Note that I am talking about the _current_ code when I say it does not fail.  Obviously it fails taking the PR into consideration because you've coded it to do so.  

Based on this line "Also try changing line 411 to say "return" rather than "go to exit_script".", I'm wondering now if your original comment was meant to indicate an internal dialogue you had at an intermediate step (i.e. you coded it to not go to exit_script, and then you were commenting on that version of the code, hence changing it to exit_script).  This is not clear at all from the comment.  Instead, it sounds like you added the block of code in question because in its current state, "It does not return a value that the consumer could check and it does not stop further execution when there is a problem. It just logs an error message and trudges on."

> It definitely should not do that in the case of an unauthenticated group1 user. 

This is an opinion again (one that I disagree with).

> Otherwise there will just be scads of messages saying "The CCL session must be authenticated to execute this test" followed by scads of messages saying "cclutAssertMockTableIsCreated(XXXX) - Table has not been created." followed by whatever assert failures occur because of those problems. The resolution is to go to exit_script rather than return in that case.

Reading this under my new assumption that this was an internal dialogue about an intermediate state of the code seems to make more sense.  Once again, that was still not immediately evident.  This comment seemed to imply that the current state would fail to create tables (which it does not).  The last line (under my original reading) made it sound like we should go to exit_script instead of completing the table creation and returning control to the user.  I'm now of the belief that "return" here meant that you were always intending on creating a message at this spot, and you were talking about returning at that point (prior to table creation) vs. going to exit_script.

My argument is to not even have this block at all and allow table creation to happen.  That avoids the original scenario you're describing of an assortment of table creation errors.

> Do you agree that the framework should not create tables when it knows it cannot remove them?
Or do you think it should go ahead and create whatever tables the consumer asked for and then message out "Unable to drop table. Manual cleanup may be required. Mock table name is X" for every single one of them and leave it to the consumer to do said manual cleanup?

Assuming infinite time/money/resources/etc., this would be my order of preference:
1. Create the tables that the consumer asks for, run the tests, attempt to clean it up, if it is not possible, present some message to the user that is reasonable for them to notice (in this case, I'm assuming it would be something in the test_results output), but still pass the test.  If this is a EOD, container, or some other place where it is of little importance that they be cleaned up, then great, the user doesn't need to do anything more, and their tests pass.

2. Create the tables that the consumer asks for, run the tests, attempt to clean it up, if it is not possible, present an error so that the user can clean it up.  They can still technically see that their assertions passed, but this makes the issue more visible (as an error) that there was a failure to clean up the tables in case they're in a domain where it matters, and then they can do so using the names that were output (note that they only need to remove the CCL definitions as the RDB entries are successfully removed).  This is the current code.

3. Throw an error from cclutCreateMockTable but move on to the next test (this is the suggestion you were leaning towards when you said "it could certainly make sense to fail fast in the current test (not the entire test case), if there were a way to pull that off.").  I have not given it a ton of thought because I still don't like this option, but there may be some way to have a #next_test label outside the loop to move on to the next test.  We'd probably need to confirm the behavior for setup/teardowns in that situation to make sure it works as expected, but I assume something could be done there.

4. The current PR.

For your final comment (nested comments aren't a thing in GitHub, so I don't want to confuse the matter more), you made two statements.

> Yes. I had done that manually, but forgot I hadn't made a unit test for it (essentially the example from the issue).

This one made sense to me.

> Another thing that should happen is forcing the session to be authenticated (for group 1 users). It doesn't really need to occur unless cclutCreateMockTable gets called.

This one did not, and I commented (on the other PR which you've quoted at the top here) as such:
> I know you mentioned it in the conversation for #29 but I wasn't really sure what you were talking about, and since it didn't impact the review, I set it aside.

Naturally, I do now (or at least I'm fairly confident assuming the _actual_ issue is tdbexecute's failure to work when unauthenticated, and assuming your original comments were about a state of code where you were erroring and returning prior to table creation), but I still disagree with these changes.

Reply:
I apologize for initiating this discussion in the context of PR #31. That was a mistake and that's why I moved it to this precise location which is where it belongs. 

I apologize for thinking you eagerly inspect things I push to this repository. Your past behavior gave me that impression. This discussion makes more sense with my last push in mind.

My last push added code causing cclutCreateMockTable to produce an error and stop further processing when the user is not a DBA and the session is not authenticated. That happens in the block containing line 411 (about one half inch south of this comment thread when viewed in context).

The best way to appreciate what I am talking about when I say "scads of messages" and "amateurish" is to download and install the version of cclutmock_table.inc from this PR after changing line 411 to **return**.  As for a concrete example, try ut_ut_cclut_emh.inc from src/test/resources. Note that the VC variables executeParams and executeNamespace must be declared for that example to run.  

I did not intend to imply that the current messaging is amateurish. I do apologize if you got that impression. I was talking about the messaging generated by **this version** of the code when **return** is used instead of **go to exit_script** at line 411.


Out of respect for people's time, I disagree about going ahead and creating every requested mock table (probably lots of them) only to fail the test case with one error message per mock table saying "couldn't clean up table XYZ. you need to do it" rather than presenting as quickly as possible one message saying "log in before running this test case" which the user will have to do in order for the test case to pass anyways.

I realize the "must login" message could come after messages from problematic tests that do not mock tables, If the user chooses to focus on fixing those things and doesn't see the "must login" message. I say no harm done. No garbage was created and no time was wasted.

Reply:
> I apologize for thinking you eagerly inspect things I push to this repository. Your past behavior gave me that impression. This discussion makes more sense with my last push in mind.

I don't intend for this to sound like criticism, but it is the case that I do this (at least to the best of my ability), and I did so here.  I believe my confusion came from the fact that I've only see two versions of this code.  The current state and the PR (which generated an error and failed fast).  I've always (from my first comment) been aware of lines 409-412 in this commit, but I have never seen a version (there is no commit in this PR's history) that involved generating an error and immediately returning (i.e. _not_ going to exit_script).  I assume that was something you tried and were unsatisfied with, which led to this comment:

> In the process, I found that perhaps cclutCreateMockTable does not work as one would like.

From there, you were commenting on a version of the code that generated an error and immediately returned.  I began assuming the comments from there were referring to the current version, and that's why this new block was added.  My confusion was only compounded because I never understood your original comment from June 7:

> Another thing that should happen is forcing the session to be authenticated (for group 1 users). It doesn't really need to occur unless cclutCreateMockTable gets called.

I didn't understand why it would need to be forced, and I never really pieced it together, likely because I didn't (and still don't) believe that it should be stopping people anyway.  (Naturally, now, I understand the comment.)

> The best way to appreciate what I am talking about when I say "scads of messages" and "amateurish" is to download and install the version of cclutmock_table.inc from this PR after changing line 411 to return. As for a concrete example, try ut_ut_cclut_emh.inc from src/test/resources. Note that the VC variables executeParams and executeNamespace must be declared for that example to run.

So, to be fully clear, I 100% agree that going to exit_script is better than immediately returning.  If I were ranking it with the other 4, immediately returning would be dead last at 5.  If that is all you're trying to convince me of, then yes, you'll get no argument from me there.

> Out of respect for people's time, I disagree about going ahead and creating every requested mock table (probably lots of them) only to fail the test case with one error message per mock table saying "couldn't clean up table XYZ. you need to do it" rather than presenting as quickly as possible one message saying "log in before running this test case" which the user will have to do in order for the test case to pass anyways.

This is a very unusual scenario to begin with because the user is logging into a back-end session and explicitly not authenticating (DVDev or the maven plugin would've already been authenticated).  In this case, the user is likely more savvy than the average CCL user, and may even understand what they're doing.  If they want the cclut program itself to say there are no errors, then yes, they need to login.  If they're just checking that their assertions pass, then they wouldn't need to login.

In my mind, this is more similar to a file not being cleaned up than a failed assertion.  If the framework tried to remove one of its files, but was unable to do so (regardless of the reason), I would want to be aware of it as a user, but I wouldn't expect it to say that my tests failed.  The error is more to let the user know what has happened than it is to inform them that something is wrong with their tests.  If that means that the wording of the messages needs to be changed or documentation to be updated, so be it.

As I said before, I'm not sure what you're looking for from me.  I still disagree (I haven't seen anything to change my mind), but since you're the maintainer, I can't stop you from merging it.  My answer to the original question:

> Do we want it to do the same for the other kinds of problems that can occur?

is still "no".

Reply:
I did not sense criticism, just misunderstanding. I was talking about the code in this PR and you were talking about something else
>This is false. No error message is logged. It "trudges on" because the creation succeeded. 
 Creating a table does not fail in an unauthenticated group1 session.  

....not with this PR. 


Even before this PR, cclutCreatMockTable can fail to create a mock table as requested yet the framework simply trudges on when that happens.  See lines 416, 422 and 425. The latter two do not create errors, but probably they should. If those lines execute, then there is a mistake in the test code. It may or may not have an impact on the test's objective, but if it does, error messaging could make it easier to identify the matter.

> I assume that was something you tried and were unsatisfied with, which led to this comment:

Correct, but I agree the answer to the question is "no".  On the other hand I think exiting the current unit test would be appropriate in at least some cases (when a request to create a mock table does not get fulfilled for example).  I just don't know how to do it.  I don't know how to exit the current unit test without also exiting the test case.

Reply:
> I was talking about the code in this PR and you were talking about something else

This is also false.  I was always talking about the new code that you added in this PR compared to the existing code (as comments in the PR should discuss).  You were talking about some intermediate state which you never fully explained, and I had to deduce through more of your comments as you made them.

> Even before this PR, cclutCreatMockTable can fail to create a mock table as requested yet the framework simply trudges on when that happens. See lines 416, 422 and 425. The latter two do not create errors, but probably they should. If those lines execute, then there is a mistake in the test code. It may or may not have an impact on the test's objective, but if it does, error messaging could make it easier to identify the matter.

The purpose of my comment was to respond to your comment that indicated an unauthenticated group1 session would log an error message and trudge on.  That is false.  It does not do that.  It does not do that in the existing code (where it will successfully create the table and execute the tests) and it does not do that in this PR (it exits everything, so there is no trudging on).

The lines you call out are irrelevant to a discussion about unauthenticated group1 sessions.  You're more than welcome to modify those lines with messaging.

> I just don't know how to do it. I don't know how to exit the current unit test without also exiting the test case.

Once again, I haven't given it a ton of thought because I don't like the option (though I do prefer it over the current PR), but I would think adding a #next_test label possibly here:
https://github.com/cerner/cclunit-framework/blob/50f3201450744db26bb9bb2a6477ff529ae7efe3/src/main/ccl/cclut_execute_test_logic.inc#L274

And then calling "go to next_test" would suffice.

Reply:
except that labels are not permitted within loops or subroutines. Creating a program to execute a test subroutine and Moving the execution of the test subroutine I suppose the logic to perform the testing could be moved to a program

Reply:
As mentioned, I haven't given a ton of thought to this, but the loop problem is trivial because the for loop can be changed to a while with the label being moved prior to that.  The cclutRegistryTestIndex would need to be incremented at that point as well.

And while executeTestLogic is its own subroutine, a cursory look would appear to indicate that it's only in a subroutine for the sake of ordered grouping.  Other advantages (such as reusability or testability) don't appear to be at play, so it could likely be embedded straight into its calling location.  There's still the conditionals to deal with, but those can also be managed.  It's certainly not impossible if the desire is strong to have the ability within the framework to short-circuit a test.

Reply:
Apparently this is a terrible forum for this this type of discussion.

>The purpose of my comment was to respond to your comment that indicated an unauthenticated group1 session would log an error message and trudge on. 

I never made that statement. What I did say is that other parts of the code do behave that way, i.e., trudge on in error situations (the exact lines noted above) and that the \[new\] code handing unauthenticated, non-DBA users should definitely not behave that way for if it did behave that way the output would be a mess. 

My exact statement is slightly incorrect (misleading?) because it says "logs an error message an\[d\] trudges on" when, even worse, there isn't even an error message in some cases.

Reply:
> I never made that statement.

I know.  I had to make assumptions because the original intention was unclear.

> What I did say is that other parts of the code do behave that way, i.e., trudge on in error situations (the exact lines noted above)

The exact lines were not noted in the original comment.  Given that the comment was about how unauthenticated group1 sessions should _not_ behave that way (as you mention), I was left to assume that you meant unauthenticated group1 sessions _would_ behave that way if the new code was not added (as there was no other code referenced or present where such a situation would occur).

Reply:
>the loop problem is trivial 

Here is the exact goal I was contemplating:  Provide the ability to exit the current unit test without exiting the current test case (i.e., continuing on to the next unit test) while still reporting any errors and assert outcomes that did get generated within the test. Aside: Currently, assert outcomes are not reported when errors occur, but that is an unrelated matter.

What is the point? It supports defensive coding in the framework that stops further processing of a single test when something really bad happens. Any additional processing will likely be invalid from that point forward. It allows the error log to end on the error that matters rather than continuing on with noise from untold many errors that will mostly just be side effects from the initiating error. 

It is not that trivial.

The test case execution is currently performed within in a loop and CCL does not allow labels in loops (of any kind). An initial thought might be to move the line **call parser(concat("call ", cclut::testName, "(null) go"))** into a subroutine and call that subroutine instead, but CCL does not allow labels in subroutines either.  Moving "that line" into a program (cclut_unit_test_executor, say) and changing the line to **execute cclut_unit_test_executor cclut::testName** would work.

I agree this is a trivial change, but the logic behind it is non-trivial.


As you hinted, the goal could be achieved without a support program, but as far as I can tell that would require spaghetti code to manually craft the loop which allows each unit test to be executed, something along the lines of 
```
#start_unit_test
;.... 
;all parser(concat("call ", cclut::testName, "(null) go")
;....
#end_unit_test
set testNumber = testNumber + 1
if (testNumber < testCount)
    go to start_unit_test
endif
```
with code saying `go to end_unit_test` whenever warranted.

Reply:
The overall logic (incorporating all pieces) is not trivial, I agree.  I was specifically referring to resolving the "label-in-a-loop" problem, which I do believe that specific problem is trivial to resolve (and the implementation you provide is one such way to do it).  Yes, you could also use the program as well (I wasn't insinuating that the label approach was the only way to solve it).

CCL doesn't have a "throwing" construct of exceptions (likely due to its nature as a scripting language), which makes the problem more complex.  However, I wouldn't call your proposed solution spaghetti code.  It's still very structured with little movement (the movement being to supplement the lack of a throwing construct).

Now, if you had 10 labels all bouncing around or labels within labels, yes, that would be quite unwieldy and program flow would be difficult to read and unpredictable without further dissection.  As it stands, it's just emulating a loop.